### ISSUE TITLE
Bus error and crashes due to memory exhaustion on one NUMA node

### ISSUE THREAD

**Issue by angainor** [openucx/ucx#7861]
With the newest HPCX 2.10.0 we see problems that very often result in user MPI applications crashing with 
```
Caught signal 7 (Bus error: nonexistent physical address)
```
This is reproducible, we now have a simple way to trigger the above issue, and it seems to me it is related to UCX. The following tests have been performed on 32 compute nodes (128 ranks per node) running MLNX_OFED_LINUX-5.5-1.0.3.2, but the same holds for MLNX_OFED_LINUX-5.4-3.1.0.0. HPCX 2.10.0 is used.

To briefly summarize:

 - when using DC transport, a large amount of memory is allocated on first NUMA nodes, probably by UCX. The memory (32 GB) is exhausted on two first NUMA nodes for already 32 compute nodes with the below described OSU alltoall test.
 
 - when memory on NUMA node 0 is exhausted in any way (like the above, or by user application), the application ultimately crashes with Bus error.

One simple way to reproduce the problem is to run the following on 32 compute nodes (OpenMPI 4.1.1):
```
mpirun -x UCX_TLS=dc,sm,self ./osu_alltoall -f -m 65536:
```
Note that I explicitly set the UCX_TLS variables: this is important, as with the OSU benchmark the problems are triggered with DC and RC transports. When I run the above, the code crashes with the following error:

```
[b2202:101358:0:101358] Caught signal 7 (Bus error: nonexistent physical address)

/cluster/software/hpcx/2.10/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so: [ ???() ]
        1 ELF        2         3 
==== backtrace (tid: 101358) ====
 0 0x0000000000007540 ???()  /cluster/software/hpcx/2.10/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so:0
 1 0x000000000006dbad ucx_recv_nb()  bcol_ucx_p2p_alltoall_pairwise.c:0
 2 0x000000000006e9fe hmca_bcol_ucx_p2p_alltoall_pairwise_progress()  ???:0
 3 0x000000000006a898 hmca_bcol_ucx_p2p_alltoall_tuned_progress()  ???:0
 4 0x0000000000073fd0 hmca_coll_ml_alltoall()  ???:0
 5 0x0000000000008c6b mca_coll_hcoll_alltoall()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/hcoll/coll_hcoll_ops.c:317
 6 0x0000000000068a16 PMPI_Alltoall()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mpi/c/profile/palltoall.c:110
 7 0x00000000004019a1 main()  /cluster/home/marcink/src/osu-micro-benchmarks-5.7/mpi/collective/osu_alltoall.c:100
 8 0x0000000000022545 __libc_start_main()  ???:0
 9 0x0000000000401d56 _start()  ???:0
=================================
[b2202:101358] *** Process received signal ***
[b2202:101358] Signal: Bus error (7)
[b2202:101358] Signal code:  (-6)
[b2202:101358] Failing at address: 0xc8fd00018bee
[b2202:101358] [ 0] /usr/lib64/libpthread.so.0(+0xf630)[0x7f9324d20630]
[b2202:101358] [ 1] /cluster/software/hpcx/2.10/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so(+0x7540)[0x7f931c2ff540]
[b2202:101358] [ 2] /cluster/software/hpcx/2.10/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so(+0x6dbad)[0x7f931c365bad]
[b2202:101358] [ 3] /cluster/software/hpcx/2.10/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so(hmca_bcol_ucx_p2p_alltoall_pairwise_progress+0x9ce)[0x7f931c3669fe]
[b2202:101358] [ 4] /cluster/software/hpcx/2.10/hcoll/lib/hcoll/hmca_bcol_ucx_p2p.so(hmca_bcol_ucx_p2p_alltoall_tuned_progress+0x148)[0x7f931c362898]
[b2202:101358] [ 5] /cluster/software/hpcx/2.10/hcoll/lib/libhcoll.so.1(hmca_coll_ml_alltoall+0xf20)[0x7f931dd47fd0]
[b2202:101358] [ 6] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/openmpi/mca_coll_hcoll.so(mca_coll_hcoll_alltoall+0x18b)[0x7f931e012c6b]
[b2202:101358] [ 7] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(MPI_Alltoall+0x36)[0x7f93255caa16]
[b2202:101358] [ 8] ./osu_alltoall[0x4019a1]
[b2202:101358] [ 9] /usr/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f9324965545]
[b2202:101358] [10] ./osu_alltoall[0x401d56]
[b2202:101358] *** End of error message ***
```

A similar error happens when I turn off HCOLL, so it can't be an HCOLL problem only. Maybe it's deeper, in UCX, or MOFED?

```
mpirun -x UCX_TLS=dc,sm,self -mca coll_hcoll_enable 0 ./osu_alltoall -f -m 65536:

[b2231:114927:0:114927] Caught signal 7 (Bus error: nonexistent physical address)
==== backtrace (tid: 114927) ====
 0 0x0000000000003d80 mca_pml_ucx_del_procs()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/pml/ucx/pml_ucx.c:487
 1 0x000000000003151c opal_progress()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/opal/runtime/opal_progress.c:231
 2 0x0000000000055cfa sync_wait_st()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/../opal/threads/wait_sync.h:83
 3 0x0000000000055cfa ompi_request_default_wait_any()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/request/req_wait.c:133
 4 0x00000000000974a1 ompi_coll_base_alltoall_intra_linear_sync()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/base/coll_base_alltoall.c:438
 5 0x0000000000005ec1 ompi_coll_tuned_alltoall_intra_dec_fixed()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:407
 6 0x0000000000068a16 PMPI_Alltoall()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mpi/c/profile/palltoall.c:110
 7 0x00000000004019a1 main()  /cluster/home/marcink/src/osu-micro-benchmarks-5.7/mpi/collective/osu_alltoall.c:100
 8 0x0000000000022545 __libc_start_main()  ???:0
 9 0x0000000000401d56 _start()  ???:0
=================================
[b2231:114927] *** Process received signal ***
[b2231:114927] Signal: Bus error (7)
[b2231:114927] Signal code:  (-6)
[b2231:114927] Failing at address: 0xc8fd0001c0ef
[b2231:114927] [ 0] /usr/lib64/libpthread.so.0(+0xf630)[0x7fae047a1630]
[b2231:114927] [ 1] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/openmpi/mca_pml_ucx.so(mca_pml_ucx_progress+0x0)[0x7fadff1dad80]
[b2231:114927] [ 2] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libopen-pal.so.40(opal_progress+0x2c)[0x7fae0434151c]
[b2231:114927] [ 3] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(ompi_request_default_wait_any+0x1aa)[0x7fae05038cfa]
[b2231:114927] [ 4] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(ompi_coll_base_alltoall_intra_linear_sync+0x321)[0x7fae0507a4a1]
[b2231:114927] [ 5] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/openmpi/mca_coll_tuned.so(ompi_coll_tuned_alltoall_intra_dec_fixed+0x81)[0x7fadfd2cdec1]
[b2231:114927] [ 6] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(MPI_Alltoall+0x36)[0x7fae0504ba16]
[b2231:114927] [ 7] ./osu_alltoall[0x4019a1]
[b2231:114927] [ 8] /usr/lib64/libc.so.6(__libc_start_main+0xf5)[0x7fae043e6545]
[b2231:114927] [ 9] ./osu_alltoall[0x401d56]
[b2231:114927] *** End of error message ***
--------------------------------------------------------------------------
```

We have diagnosed the issue and it seems that the crash is caused by memory exhaustion on some numa nodes. When run, the above code eats up the memory on NUMA node 0, then continues allocating from NUMA node 1, as can be observed during the run by 
```
clush -w <nodelist> numactl -H | grep free
```
Right before the bus error, this is what we see:
```
b2236: node 0 free: 288 MB
b2236: node 1 free: 827 MB
b2236: node 2 free: 13214 MB
b2236: node 3 free: 12489 MB
b2236: node 4 free: 13379 MB
b2236: node 5 free: 13421 MB
b2236: node 6 free: 13411 MB
b2236: node 7 free: 13361 MB
```

We have now changed the default UCX transport on our system to UCX_TLS=ud,sm,self. This fixes the problem that you see above, i.e., OpenMPI/UCX/HCOLL no longer steal all of the RAM from the first NUMA nodes, the NUMA usage is balanced, and the OSU benchmark runs through.

However, this does not solve the Bus error problem in general: if the user application uses up all ram on NUMA node 0, the code still crashes with Bus error. You can verify that by running a simple memory allocator on NUMA Node 0:
```
numactl --physcpubind=0 ./memtest
```
where memtest.c is
```
#include <stdlib.h>
#include <string.h>

int main()
{
    void *ptr;
    ptr = malloc((size_t)1024*1024*1024*32);
    memset(ptr, 1, (size_t)1024*1024*1024*32);
    while(1);
}
```
If I do that, the OSU benchmark crashes with Bus error also for the UD transport.

Is there anything that comes to your mind that could cause the issue? Does it look like a UCX issue, or is it caused by something else?

**Comment by yosefe** [comment]
@angainor if the user application uses up all ram on NUMA node 0, and OSU benchmark is run without UCX (OpenMPI TCP for example), does it still crash?

**Comment by angainor** [comment]
@yosefe Thank you! I did not think about that test. I run
```
mpirun -mca btl tcp,vader,self -mca pml ^ucx -mca coll_hcoll_enable 0 ./osu_alltoall -f -m 65536:
```
and it crashes with the following:
```
bash-4.2$ mpirun  -mca btl tcp,vader,self -mca pml ^ucx -mca coll_hcoll_enable 0 ./osu_alltoall -f -m 65536:
[b2238:126065:0:126065] Caught signal 7 (Bus error: nonexistent physical address)
==== backtrace (tid: 126065) ====
 0 0x0000000000055a45 ompi_request_wait_completion()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/../ompi/request/request.h:415
 1 0x0000000000055a45 ompi_request_default_wait()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/request/req_wait.c:42
 2 0x000000000009c1f4 ompi_coll_base_sendrecv_zero()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/base/coll_base_barrier.c:64
 3 0x000000000009c1f4 ompi_coll_base_barrier_intra_bruck()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/base/coll_base_barrier.c:271
 4 0x0000000000069328 PMPI_Barrier()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mpi/c/profile/pbarrier.c:74
 5 0x000000000040193e main()  /cluster/home/marcink/src/osu-micro-benchmarks-5.7/mpi/collective/osu_alltoall.c:95
 6 0x0000000000022545 __libc_start_main()  ???:0
 7 0x0000000000401d56 _start()  ???:0
=================================
[b2238:126065] *** Process received signal ***
[b2238:126065] Signal: Bus error (7)
[b2238:126065] Signal code:  (-6)
[b2238:126065] Failing at address: 0xc8fd0001ec71
[b2238:126065] [ 0] /usr/lib64/libpthread.so.0(+0xf630)[0x7f8cc3b44630]
[b2238:126065] [ 1] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(ompi_request_default_wait+0x105)[0x7f8cc43dba45]
[b2238:126065] [ 2] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(ompi_coll_base_barrier_intra_bruck+0xa4)[0x7f8cc44221f4]
[b2238:126065] [ 3] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(MPI_Barrier+0x38)[0x7f8cc43ef328]
[b2238:126065] [ 4] ./osu_alltoall[0x40193e]
[b2238:126065] [ 5] /usr/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f8cc3789545]
[b2238:126065] [ 6] ./osu_alltoall[0x401d56]
[b2238:126065] *** End of error message ***
```
I guess it's deeper than I thought.. Any ideas what that could be?

**Comment by angainor** [comment]
Although to be honest, I am not sure that the TCP transport works as it's supposed to. `osu_barrier` hangs forever. I think there is a connection issue. But the Bus error still happened.

**Comment by yosefe** [comment]
i would guess it has something to do with memory binding that mpirun might be doing. for the TCP connection issues, can try to use "-mca btl_tcp_if_include " filter.

**Comment by angainor** [comment]
@yosefe yes, `-mca btl_tcp_if_include ib0` works. The bus error is still triggered:
```
bash-4.2$ mpirun -mca btl_tcp_if_include ib0 -mca mca_btl_base_verbose 100 -mca btl tcp,vader,self -mca pml ^ucx -mca coll_hcoll_enable 0 ./osu_alltoall -f

# OSU MPI All-to-All Personalized Exchange Latency Test v5.7
# Size       Avg Latency(us)   Min Latency(us)   Max Latency(us)  Iterations
1                    1207.72           1095.63           1434.07        1000
2                    1899.03           1760.52           2075.24        1000
4                    2749.48           2484.31           2986.79        1000
8                    5004.98           4422.62           5452.30        1000
16                   9399.63           7811.73          10463.53        1000
32                  19485.22          15495.25          22405.92        1000
[b2234:123960:0:123960] Caught signal 7 (Bus error: nonexistent physical address)
==== backtrace (tid: 123960) ====
 0 0x00000000000052c0 mca_btl_vader_poll_handle_frag()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/opal/mca/btl/vader/btl_vader_frag.h:148
 1 0x000000000003151c opal_progress()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/opal/runtime/opal_progress.c:231
 2 0x0000000000055a45 ompi_request_wait_completion()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/../ompi/request/request.h:415
 3 0x0000000000055a45 ompi_request_default_wait()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/request/req_wait.c:42
 4 0x0000000000092b1a ompi_coll_base_sendrecv_actual()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/base/coll_base_util.c:62
 5 0x0000000000096fee ompi_coll_base_sendrecv()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/base/coll_base_util.h:133
 6 0x0000000000096fee ompi_coll_base_alltoall_intra_bruck()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/base/coll_base_alltoall.c:274
 7 0x0000000000005ec1 ompi_coll_tuned_alltoall_intra_dec_fixed()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mca/coll/tuned/coll_tuned_decision_fixed.c:407
 8 0x0000000000068a16 PMPI_Alltoall()  /cluster/work/users/vegarde/build/OpenMPI/4.1.1/GCC-11.2.0/openmpi-4.1.1/ompi/mpi/c/profile/palltoall.c:110
 9 0x00000000004019a1 main()  /cluster/home/marcink/src/osu-micro-benchmarks-5.7/mpi/collective/osu_alltoall.c:100
10 0x0000000000022545 __libc_start_main()  ???:0
11 0x0000000000401d56 _start()  ???:0
=================================
[b2234:123960] *** Process received signal ***
[b2234:123960] Signal: Bus error (7)
[b2234:123960] Signal code:  (-6)
[b2234:123960] Failing at address: 0xc8fd0001e438
[b2234:123960] [ 0] /usr/lib64/libpthread.so.0(+0xf630)[0x7f6ff80d2630]
[b2234:123960] [ 1] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/openmpi/mca_btl_vader.so(+0x52c0)[0x7f6ff2b2b2c0]
[b2234:123960] [ 2] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libopen-pal.so.40(opal_progress+0x2c)[0x7f6ff7c7251c]
[b2234:123960] [ 3] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(ompi_request_default_wait+0x105)[0x7f6ff8969a45]
[b2234:123960] [ 4] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(ompi_coll_base_sendrecv_actual+0xaa)[0x7f6ff89a6b1a]
[b2234:123960] [ 5] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(ompi_coll_base_alltoall_intra_bruck+0x4de)[0x7f6ff89aafee]
[b2234:123960] [ 6] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/openmpi/mca_coll_tuned.so(ompi_coll_tuned_alltoall_intra_dec_fixed+0x81)[0x7f6ff146fec1]
[b2234:123960] [ 7] /cluster/software/OpenMPI/4.1.1-GCC-11.2.0/lib/libmpi.so.40(MPI_Alltoall+0x36)[0x7f6ff897ca16]
[b2234:123960] [ 8] ./osu_alltoall[0x4019a1]
[b2234:123960] [ 9] /usr/lib64/libc.so.6(__libc_start_main+0xf5)[0x7f6ff7d17545]
[b2234:123960] [10] ./osu_alltoall[0x401d56]
[b2234:123960] *** End of error message ***
--------------------------------------------------------------------------
```

**Comment by angainor** [comment]
@yosefe The binding has not changed, and this crash only showed up after our December upgrade to MLNX_OFED_LINUX-5.4-3.1.0.0, new Lustre client, and HPCX 2.10.0. I don't think user-space application should be able to cause this. Don't you think this must come from the kernel somewhere? I mean, whatever the binding is, the default Linux allocation strategy is that if the code runs out of RAM on one numa node, it takes from another, where there is free memory. We do not force local node only.

In either case, I guess this is not related to UCX after all, since the TCP transport crashes in the same way. Thanks again for the test!

**Comment by angainor** [comment]
@yosefe and the second question remains: why does UCX allocate all of RAM from NUMA node 0 and 1 when I use the DC transport? Do you think I can somehow diagnose that better?

**Comment by yosefe** [comment]
> @yosefe and the second question remains: why does UCX allocate all of RAM from NUMA node 0 and 1 when I use the DC transport? Do you think I can somehow diagnose that better?

Can you pls capture the following:
1. Set `UCX_HANDLE_ERRORS=freeze` to hold the process in bad state

Then, when it fails and freezes:
1. `for p in $(pgrep test_reg_mr); do echo $p; cat /proc/$p/numa_maps; done > numa_maps.txt` on one failing node
1. `cat /sys/class/infiniband/*/fw_pages > fw_pages.txt`
1. `odp_stat.sh > odp_stat.log`
1. Upload the 3 logs above

**Comment by angainor** [comment]
@yosefe The first line does not produce any output. `pgrep test_reg_mr` is empty.

2.
```
[root@b2202.BETZY ~]# cat /sys/class/infiniband/*/fw_pages > fw_pages.txt
[root@b2202.BETZY ~]# cat fw_pages.txt 
9045694
```

3.
```
[root@b2202.BETZY ~]# odp_stat.sh > odp_stat.log
[root@b2202.BETZY ~]# cat odp_stat.log 
mlx5_0
type			pages		count
odp regions		NA		NA
odp page faults		NA		NA
odp invalidations	NA		NA
odp prefetches		NA		NA
active prefetches			NA
odp fault contentions			NA
odp faults wrong key			NA
odp unhandled faults			NA
			cur	size	limit	miss
mr cache 16K		1499	607440	750	678593
mr cache 32K		1499	45685	750	374
mr cache 64K		499	33118	250	1805
mr cache 128K		499	31879	250	0
mr cache 256K		499	29310	250	0
mr cache 512K		499	28735	250	0
mr cache 1.0M		499	28016	250	0
mr cache 2.0M		499	27014	250	0
mr cache 4.0M		499	26439	250	0
mr cache 8.0M		499	26094	250	0
mr cache 16M		499	25749	250	0
mr cache 32M		499	25362	250	0
mr cache 64M		63	22244	32	0
mr cache 128M		31	21470	16	0
mr cache 256M		15	17568	8	0
mr cache 512M		7	13988	4	0
mr cache 1.0G		0	0	0	0
mr cache 2.0G		0	0	0	0
mr cache 4.0G		0	0	0	0
mr cache 8.0G		0	0	0	0
mr cache 16G		0	0	0	0
mr cache leaf		0	0	0	0
mr cache iodp		0	0	0	0
```

**Comment by angainor** [comment]
@yosefe I guess you wanted the maps for a crashed process? This is from one that sent the Bus fault.

[numa_maps.txt](https://github.com/openucx/ucx/files/7904870/numa_maps.txt)

**Comment by yosefe** [comment]
@angainor sorry about that, it should be osu_alltoall and not test_reg_mr.
can you pls collect numastat:
`for p in $(pgrep osu_alltoall); do echo $p; numastat -p $p; done > numastat.txt`

Anyway, it seems FW is consuming the NUMA0 memory, which kinda makes sense since it's not aware of which process runs on each NUMA node, and probably tries to allocate memory closest to NIC.
If we can also confirm from the previous step that it's not coming from userspace, the next step would be to open a Mellanox support case.

**Comment by angainor** [comment]
@yosefe Here is the output of the `numastat`. Note that I think this is dependent on the problem size. `osu_alltoall` triggers this for arrays of size 32k+, not for smaller ones. So it seems there are some data size-dependent communication buffers allocated here.

[numastat.txt](https://github.com/openucx/ucx/files/7904938/numastat.txt)

**Comment by yosefe** [comment]
according to this log, no process allocates more than 3MB of VM outside of its assigned NUMA node. So it must be coming from fw_pages.

**Comment by angainor** [comment]
@yosefe ok, thanks a lot! I will take this up with the vendor and Mellanox.

Thanks a lot for your help in diagnosing this, and for the `UCX_HANDLE_ERRORS=freeze` option - it's absolutely brilliant!

**Comment by stdweird** [comment]
@angainor we have been seeing this as well. can you share any updates/feedback/fixes from mellanox here when they get back to you? or if you have a case number, i can open my own case and refer then to yours (maybe to bump the prio a bit ;)

**Comment by angainor** [comment]
@stdweird we have a case through the vendor, so unfortunately I don't have direct access to Mellanox and I don't know their case number. What I've figured out is that

1. the Bus error was related to the Lustre client. We had version 2.12.6. Now that we upgraded to 2.12.8 the Bus error is gone.
2. the memory exhaustion is still not pinned, but I do know that it's somehow related to HPCX. When I run the benchmark with HPCX 2.8.1, I don't see the issue. The problem occured in version 2.9.0 and up.

Hope this helps. Let me know if you get more info about this.
